jerry-core/jcontext/jcontext.h

```diff
@@ -116,6 +117,77 @@ typedef struct
 } jerry_context_t;
 
 /**
+ * Description of jerry instance, which contains the meta data and buffer.
+ */
+struct jerry_instance_t
+{
+  struct
+  {
+    uint32_t heap_size; /**< the heap size */
+    uint8_t *context_p; /**< point to the entrance of the context in buffer */
+    uint8_t *heap_p; /**< point to the entrance of the heap in buffer */
+    uint8_t *lcache_p; /**< point to the entrance of the lcache in buffer */
+  } head;
+  uint8_t buffer[]; /**< the flexible array for storing context, heap and lcache */
+};
+
+#ifndef CONFIG_ECMA_LCACHE_DISABLE
+/**
+ * Hash table for caching the last access of properties.
+ */
+typedef struct
+{
+  /**
+   * Hash table
+   */
+  ecma_lcache_hash_entry_t table[ECMA_LCACHE_HASH_ROWS_COUNT][ECMA_LCACHE_HASH_ROW_LENGTH];
+} jerry_hash_table_t;
+#endif /* !CONFIG_ECMA_LCACHE_DISABLE */
+
+#ifdef JERRY_ENABLE_EXTERNAL_CONTEXT
+
+/**
+ * This part is for Jerry which enable external context.
+ */
+
+typedef struct
+{
+  jmem_heap_free_t first; /**< first node in free region list */
+  uint8_t area[]; /**< heap area */
+} jmem_heap_t;
+
+jerry_context_t *jerry_context_get_current_context (void);
+
+#define JERRY_CONTEXT(field) (jerry_context_get_current_context ()->field)
+
+#ifndef JERRY_SYSTEM_ALLOCATOR
+
+jmem_heap_t *jerry_context_get_current_heap (void);
+
+#define JERRY_HEAP_CONTEXT(field) (jerry_context_get_current_heap ()->field)
+
+#undef JMEM_HEAP_SIZE
```

Comment:
Undefing a symbol is dangerous since it only take effect if this header file is included (a hidden dependency). I would prefer that if this context is selected, JMEM_HEAP_SIZE must not be defined. So replace this undef to:

    #ifdef JMEM_HEAP_SIZE
    #error "JMEM_HEAP_SIZE must not be defined if JERRY_ENABLE_EXTERNAL_CONTEXT is defined"
    #endif /* JMEM_HEAP_SIZE */